File: problem3.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>

struct ColorCount {
    std::string color;
    int count;
};

bool isPortionPossible(const std::vector<ColorCount>& data) {
    for (const auto& item : data) {
        if ((item.color == "red" && item.count > 12) ||
            (item.color == "green" && item.count > 13) ||
            (item.color == "blue" && item.count > 14)) {
            return false;
        }
    }
    return true;
}

int main() {
    std::ifstream file("yourfile.txt");
    std::string line;
    int sumOfPossibleGames = 0;

    while (std::getline(file, line)) {
        std::istringstream iss(line);
        std::string gameLabel;
        int gameNumber;
        char colon;

        iss >> gameLabel >> gameNumber >> colon; // Reading "Game", game number, and colon

        std::string segment;
        bool isGamePossible = true;

        // Process the remainder of the line
        std::string remainder;
        std::getline(iss, remainder);
        std::istringstream segmentStream(remainder);

        while (std::getline(segmentStream, segment, ';')) {
            std::istringstream colorStream(segment);
            std::vector<ColorCount> portionData;
            int count;
            std::string color;

            while (colorStream >> count >> color) {
                color.erase(color.find_last_not_of(", ") + 1); // Remove trailing comma and space
                portionData.push_back({color, count});
            }

            if (!isPortionPossible(portionData)) {
                isGamePossible = false;
                break; // Stop checking further if one portion fails
            }
        }

        if (isGamePossible) {
            sumOfPossibleGames += gameNumber;
        }
    }

    std::cout << "Sum of possible game IDs: " << sumOfPossibleGames << std::endl;

    return 0;
}

File: problem3.py
def is_portion_possible(data):
    for color, count in data:
        if (color == "red" and count > 12) or \
           (color == "green" and count > 13) or \
           (color == "blue" and count > 14):
            return False
    return True

def process_file(filename):
    sum_of_possible_games = 0

    with open(filename, 'r') as file:
        for line in file:
            parts = line.strip().split()
            game_number = int(parts[1].rstrip(':'))
            is_game_possible = True

            portions = ' '.join(parts[2:]).split(';')
            for portion in portions:
                color_data = []
                items = portion.strip().split(',')
                for item in items:
                    count, color = item.strip().split()
                    color_data.append((color, int(count)))

                if not is_portion_possible(color_data):
                    is_game_possible = False
                    break

            if is_game_possible:
                sum_of_possible_games += game_number

    return sum_of_possible_games

# Replace 'yourfile.txt' with the path to your file
sum_of_possible_games = process_file('problem3.txt')
print("Sum of possible game IDs:", sum_of_possible_games)
